File: binance4j-spot/src/main/java/com/binance4j/spot/client/SpotClient.java
package com.binance4j.spot.client;

import com.binance4j.core.client.RestClient;
import com.binance4j.core.param.Params;
import com.binance4j.core.param.TimeFrame;
import com.binance4j.spot.param.AccountParams;
import com.binance4j.spot.param.AllOCOInfoParams;
import com.binance4j.spot.param.AllOrdersParams;
import com.binance4j.spot.param.CancelOCOParams;
import com.binance4j.spot.param.CancelOpenOrdersParams;
import com.binance4j.spot.param.CancelOrderParams;
import com.binance4j.spot.param.LimitMakerOrder;
import com.binance4j.spot.param.LimitOrder;
import com.binance4j.spot.param.MarketOrder;
import com.binance4j.spot.param.MarketQuoteOrder;
import com.binance4j.spot.param.NewOCOOrderParams;
import com.binance4j.spot.param.NewOrderParams;
import com.binance4j.spot.param.OCOInfoParams;
import com.binance4j.spot.param.OpenOCOParams;
import com.binance4j.spot.param.OpenOrdersStatusParams;
import com.binance4j.spot.param.OrderCountParams;
import com.binance4j.spot.param.OrderStatusParams;
import com.binance4j.spot.param.StopLossLimitOrder;
import com.binance4j.spot.param.StopLossOrder;
import com.binance4j.spot.param.TakeProfitLimitOrder;
import com.binance4j.spot.param.TakeProfitOrder;
import com.binance4j.spot.param.TradesParams;
import com.binance4j.spot.request.CancelOCORequest;
import com.binance4j.spot.request.CancelOpenOrderRequest;
import com.binance4j.spot.request.CancelOrderRequest;
import com.binance4j.spot.request.GetAccountRequest;
import com.binance4j.spot.request.GetAllOCORequest;
import com.binance4j.spot.request.GetAllOrdersRequest;
import com.binance4j.spot.request.GetOCORequest;
import com.binance4j.spot.request.GetOpenOCORequest;
import com.binance4j.spot.request.GetOpenOrdersRequest;
import com.binance4j.spot.request.GetOrderCountRequest;
import com.binance4j.spot.request.GetOrderStatusRequest;
import com.binance4j.spot.request.GetTradesRequest;
import com.binance4j.spot.request.NewOCORequest;
import com.binance4j.spot.request.NewOrderRequest;
import com.binance4j.spot.request.NewOrderTestRequest;

/**
 * API client for the SPOT endpoints
 * 
 * @see <a href=
 *      "https://binance-docs.github.io/apidocs/spot/en/#spot-account-trade">Documentation</a>
 */
public class SpotClient extends RestClient<SpotMapping> {
	/**
	 * @param key    API public key.
	 * @param secret API secret key.
	 */
	public SpotClient(String key, String secret) {
		super(SpotMapping.class, key, secret);
	}

	/**
	 * @param key        API public key.
	 * @param secret     API secret key.
	 * @param useTestnet use testnet.
	 */
	protected SpotClient(String key, String secret, boolean useTestnet) {
		super(SpotMapping.class, key, secret, useTestnet);
	}

	/**
	 * Send in a new order.
	 * 
	 * @param params Request params.
	 * @return The request to execute.
	 * @deprecated Use specific Order.
	 */
	@Deprecated
	public NewOrderRequest newOrder(NewOrderParams params) {
		return new NewOrderRequest(service.newOrder(params.toMap()));
	}

	// ORDERS

	/**
	 * Tests a {@link MarketOrder}.
	 * 
	 * @param order the {@link MarketOrder}.
	 * @return The request to execute.
	 */
	public NewOrderRequest newOrder(MarketOrder order) {
		return new NewOrderRequest(service.newOrder(order.toMap()));
	}

	/**
	 * Tests a {@link MarketQuoteOrder}.
	 * 
	 * @param order the {@link MarketQuoteOrder}.
	 * @return The request to execute.
	 */
	public NewOrderRequest newOrder(MarketQuoteOrder order) {
		return new NewOrderRequest(service.newOrder(order.toMap()));
	}

	/**
	 * Tests a {@link LimitOrder}.
	 * 
	 * @param order the {@link LimitOrder}.
	 * @return The request to execute.
	 */
	public NewOrderRequest newOrder(LimitOrder order) {
		return new NewOrderRequest(service.newOrder(order.toMap()));
	}

	/**
	 * Tests a {@link LimitMakerOrder}.
	 * 
	 * @param order the {@link LimitMakerOrder}.
	 * @return The request to execute.
	 */
	public NewOrderRequest newOrder(LimitMakerOrder order) {
		return new NewOrderRequest(service.newOrder(order.toMap()));
	}

	/**
	 * Tests a {@link StopLossLimitOrder}.
	 * 
	 * @param order the {@link StopLossLimitOrder}.
	 * @return The request to execute.
	 */
	public NewOrderRequest newOrder(StopLossLimitOrder order) {
		return new NewOrderRequest(service.newOrder(order.toMap()));
	}

	/**
	 * Tests a {@link StopLossOrder}.
	 * 
	 * @param order the {@link StopLossOrder}.
	 * @return The request to execute.
	 */
	public NewOrderRequest newOrder(StopLossOrder order) {
		return new NewOrderRequest(service.newOrder(order.toMap()));
	}

	/**
	 * Tests a {@link TakeProfitLimitOrder}.
	 * 
	 * @param order the {@link TakeProfitLimitOrder}.
	 * @return The request to execute.
	 */
	public NewOrderRequest newOrder(TakeProfitLimitOrder order) {
		return new NewOrderRequest(service.newOrder(order.toMap()));
	}

	/**
	 * Tests a {@link TakeProfitOrder}.
	 * 
	 * @param order the {@link TakeProfitOrder}.
	 * @return The request to execute.
	 */
	public NewOrderRequest newOrder(TakeProfitOrder order) {
		return new NewOrderRequest(service.newOrder(order.toMap()));
	}

	// END ORDERS

	// TEST ORDERS

	/**
	 * Creates and validates a new order but does not send it into the matching
	 * engine.
	 * 
	 * @param params Request params.
	 * @return The request to execute.
	 * @deprecated Use specific Order.
	 */
	@Deprecated
	public NewOrderTestRequest newOrderTest(NewOrderParams params) {
		return new NewOrderTestRequest(service.newOrderTest(params.toMap()));
	}

	/**
	 * Tests a {@link MarketOrder}.
	 * 
	 * @param order the {@link MarketOrder}.
	 * @return The request to execute.
	 */
	public NewOrderTestRequest newOrderTest(MarketOrder order) {
		return new NewOrderTestRequest(service.newOrderTest(order.toMap()));
	}

	/**
	 * Tests a {@link MarketQuoteOrder}.
	 * 
	 * @param order the {@link MarketQuoteOrder}.
	 * @return The request to execute.
	 */
	public NewOrderTestRequest newOrderTest(MarketQuoteOrder order) {
		return new NewOrderTestRequest(service.newOrderTest(order.toMap()));
	}

	/**
	 * Tests a {@link LimitOrder}.
	 * 
	 * @param order the {@link LimitOrder}.
	 * @return The request to execute.
	 */
	public NewOrderTestRequest newOrderTest(LimitOrder order) {
		return new NewOrderTestRequest(service.newOrderTest(order.toMap()));
	}

	/**
	 * Tests a {@link LimitMakerOrder}.
	 * 
	 * @param order the {@link LimitMakerOrder}.
	 * @return The request to execute.
	 */
	public NewOrderTestRequest newOrderTest(LimitMakerOrder order) {
		return new NewOrderTestRequest(service.newOrderTest(order.toMap()));
	}

	/**
	 * Tests a {@link StopLossLimitOrder}.
	 * 
	 * @param order the {@link StopLossLimitOrder}.
	 * @return The request to execute.
	 */
	public NewOrderTestRequest newOrderTest(StopLossLimitOrder order) {
		return new NewOrderTestRequest(service.newOrderTest(order.toMap()));
	}

	/**
	 * Tests a {@link StopLossOrder}.
	 * 
	 * @param order the {@link StopLossOrder}.
	 * @return The request to execute.
	 */
	public NewOrderTestRequest newOrderTest(StopLossOrder order) {
		return new NewOrderTestRequest(service.newOrderTest(order.toMap()));
	}

	/**
	 * Tests a {@link TakeProfitLimitOrder}.
	 * 
	 * @param order the {@link TakeProfitLimitOrder}.
	 * @return The request to execute.
	 */
	public NewOrderTestRequest newOrderTest(TakeProfitLimitOrder order) {
		return new NewOrderTestRequest(service.newOrderTest(order.toMap()));
	}

	/**
	 * Tests a {@link TakeProfitOrder}.
	 * 
	 * @param order the {@link TakeProfitOrder}.
	 * @return The request to execute.
	 */
	public NewOrderTestRequest newOrderTest(TakeProfitOrder order) {
		return new NewOrderTestRequest(service.newOrderTest(order.toMap()));
	}

	// END TEST ORDERS

	/**
	 * Cancel an active order.
	 * 
	 * @param params Request params.
	 * @return The request to execute.
	 */
	public CancelOrderRequest cancelOrder(CancelOrderParams params) {
		return new CancelOrderRequest(service.cancelOrder(params.toMap()));
	}

	/**
	 * Cancels all active orders on a symbol. This includes OCO orders.
	 * 
	 * @param params Request params.
	 * @return The request to execute.
	 */
	public CancelOpenOrderRequest cancelOpenOrders(CancelOpenOrdersParams params) {
		return new CancelOpenOrderRequest(service.cancelOpenOrders(params.toMap()));
	}

	/**
	 * Check an order's status.
	 * <ul>
	 * <li>Either orderId or origClientOrderId must be sent.</li>
	 * <li>For some historical orders cummulative quote quantity will be &lt; 0,
	 * meaning the data is not available at this
	 * time.</li>
	 * </ul>
	 * 
	 * @param params Request params.
	 * @return The request to execute.
	 */
	public GetOrderStatusRequest getOrderStatus(OrderStatusParams params) {
		return new GetOrderStatusRequest(service.getOrderStatus(params.toMap()));
	}

	/**
	 * Get all open orders on a symbol.
	 * 
	 * @param params Request params.
	 * @return The request to execute.
	 */
	public GetOpenOrdersRequest getOpenOrders(OpenOrdersStatusParams params) {
		return new GetOpenOrdersRequest(service.getOpenOrders(params.toMap()));
	}

	/**
	 * Get all open orders.
	 * 
	 * @return The request to execute.
	 */
	public GetOpenOrdersRequest getOpenOrders() {
		OpenOrdersStatusParams params = new OpenOrdersStatusParams(null);
		return new GetOpenOrdersRequest(service.getOpenOrders(params.toMap()));
	}

	/**
	 * Get all orders on a symbol.
	 * 
	 * @param params Request params.
	 * @return The request to execute.
	 */
	public GetAllOrdersRequest getAllOrders(AllOrdersParams params) {
		return new GetAllOrdersRequest(service.getAllOrders(params.toMap()));
	}

	/**
	 * Get all orders on a symbol.
	 * 
	 * @param params    Request params.
	 * @param timeFrame Time frame.
	 * @return The request to execute.
	 */
	public GetAllOrdersRequest getAllOrders(AllOrdersParams params, TimeFrame timeFrame) {
		return new GetAllOrdersRequest(service.getAllOrders(Params.merge(params, timeFrame)));
	}

	/**
	 * Send in an OCO order.
	 * 
	 * @param params Request params.
	 * @return The request to execute.
	 */
	public NewOCORequest newOCO(NewOCOOrderParams params) {
		return new NewOCORequest(service.newOCO(params.toMap()));
	}

	/**
	 * Cancel an entire Order List. Canceling an individual leg will cancel the
	 * entire OCO
	 * 
	 * @param params Request params.
	 * @return The request to execute.
	 */
	public CancelOCORequest cancelOCO(CancelOCOParams params) {
		return new CancelOCORequest(service.cancelOCO(params.toMap()));
	}

	/**
	 * Retrieves a specific OCO based on provided optional parameters.
	 * 
	 * @param params Request params.
	 * @return The request to execute.
	 */
	public GetOCORequest getOCO(OCOInfoParams params) {
		return new GetOCORequest(service.queryOCO(params.toMap()));
	}

	/**
	 * Retrieves all OCO based on provided optional parameters.
	 * 
	 * @param params Request params.
	 * @return The request to execute.
	 */
	public GetAllOCORequest getAllOCO(AllOCOInfoParams params) {
		return new GetAllOCORequest(service.getAllOCO(params.toMap()));
	}

	/**
	 * Retrieves all OCO based on provided optional parameters.
	 * 
	 * @return The request to execute.
	 */
	public GetAllOCORequest getAllOCO() {
		return new GetAllOCORequest(service.getAllOCO(new AllOCOInfoParams(null).toMap()));
	}

	/**
	 * Retrieves all OCO based on provided optional parameters.
	 * 
	 * @param params    Request params.
	 * @param timeFrame Time frame.
	 * @return The request to execute.
	 */
	public GetAllOCORequest getAllOCO(AllOCOInfoParams params, TimeFrame timeFrame) {
		return new GetAllOCORequest(service.getAllOCO(Params.merge(params, timeFrame)));
	}

	/**
	 * Retrieves all OCO based on provided optional parameters.
	 * 
	 * @param timeFrame Time frame.
	 * @return The request to execute.
	 */
	public GetAllOCORequest getAllOCO(TimeFrame timeFrame) {
		return new GetAllOCORequest(service.getAllOCO(Params.merge(new AllOCOInfoParams(null), timeFrame)));
	}

	/**
	 * Retrieves all open OCO.
	 * 
	 * @return The request to execute.
	 */
	public GetOpenOCORequest getOpenOCO() {
		return new GetOpenOCORequest(service.getOpenOCO(new OpenOCOParams().toMap()));
	}

	/**
	 * Get current account information.
	 * 
	 * @return The request to execute.
	 */
	public GetAccountRequest getAccount() {
		return new GetAccountRequest(service.getAccount(new AccountParams().toMap()));
	}

	/**
	 * Get trades for a specific account and symbol. If fromId is set, it will get
	 * id &gt;= fromId. Otherwise most recent
	 * trades are returned.
	 * 
	 * @param params Request params.
	 * @return The request to execute.
	 */
	public GetTradesRequest getTrades(TradesParams params) {
		return new GetTradesRequest(service.getTrades(params.toMap()));
	}

	/**
	 * Get trades for a specific account and symbol. If fromId is set, it will get
	 * id &gt;= fromId. Otherwise most recent
	 * trades are returned.
	 * 
	 * @param params    Request params.
	 * @param timeFrame Time frame.
	 * @return The request to execute.
	 */
	public GetTradesRequest getTrades(TradesParams params, TimeFrame timeFrame) {
		return new GetTradesRequest(service.getTrades(Params.merge(params, timeFrame)));
	}

	/**
	 * Displays the user's current order count usage for all intervals with default
	 * request
	 * 
	 * @return The request to execute.
	 */
	public GetOrderCountRequest getOrderCount() {
		return new GetOrderCountRequest(service.getOrderCount(new OrderCountParams().toMap()));
	}
}

File: src/main/kotlin/com/binance4j/connectors/spot/client/SpotClient.kt
/*
 * MIT License
 *
 * Copyright (c) 2022 Binance4j
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NON INFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
package com.binance4j.connectors.spot.client

import com.binance4j.connectors.core.Request
import com.binance4j.connectors.core.client.RestClient
import com.binance4j.connectors.core.dto.NewOrderResponseType
import com.binance4j.connectors.core.dto.OrderSide
import com.binance4j.connectors.core.dto.TimeInForce
import com.binance4j.connectors.spot.dto.NewOrder
import com.binance4j.connectors.spot.dto.OrderInfo

/**
 * API client for the SPOT endpoints
 *
 * [Documentation](https://binance-docs.github.io/apidocs/spot/en/#spot-account-trade)
 */
object SpotClient : RestClient<SpotMapping>(SpotMapping::class.java) {
    /**
     * Defines if the client uses the test network.
     *
     * @param testnet network choice.
     */
    fun testnet(testnet: Boolean): SpotClient {
        createService(SpotMapping::class.java, testnet)
        return this
    }

    /**
     * Sends an order.
     *
     * @param order The order to send.
     */
    fun newOrder(order: NewOrder) = Request(service.newOrder(order.toMap()))

    /**
     * Sends an order.
     *
     * @param order The order to send.
     */
    fun newOrderTest(order: NewOrder) = Request(service.newOrderTest(order.toMap()))

    /**
     * Cancel an active order.
     *
     * @param symbol            Trade symbol.
     * @param orderId           Order id.
     * @param origClientOrderId Original client order id.
     * @param newClientOrderId  New client order id.
     *
     * @return The request to execute.
     */
    @JvmOverloads
    fun cancelOrder(symbol: String, orderId: Long, origClientOrderId: String? = null, newClientOrderId: String? = null) =
        Request(service.cancelOrder(symbol, orderId, origClientOrderId, newClientOrderId))

    /**
     * Cancels all active orders on a symbol. This includes OCO orders.
     *
     * @param symbol Trade symbol.
     *
     * @return The request to execute.
     */

    fun cancelOpenOrders(symbol: String) = Request(service.cancelOpenOrders(symbol))

    /**
     * Check an order's status.
     *
     * Either orderId or origClientOrderId must be sent.
     * For some historical orders cumulative quote quantity will be &lt; 0, meaning the data is not available at this time.
     *

     * @param symbol            Trading pair.
     * @param origClientOrderId Client id.
     * @param orderId           Order id.
     *
     * @return The request to execute.
     */
    @JvmOverloads
    fun getOrderStatus(symbol: String, origClientOrderId: String? = null, orderId: Long? = null): Request<OrderInfo> =
        Request(service.getOrderStatus(symbol, origClientOrderId, orderId))

    /**
     * Get all open orders on a symbol.
     *
     * @param symbol Pair we want the open orders.
     *
     * @return The request to execute.
     */
    @JvmOverloads
    fun getOpenOrders(symbol: String? = null) = Request(service.getOpenOrders(symbol))

    /**
     * Get all orders on a symbol.
     *
     * @param symbol  Trading pair we want the orders.
     * @param orderId If orderId is set, it will get orders &gt;= that orderId. Otherwise, most recent orders are returned.
     * If startTime and/or endTime provided, orderId is not required.
     * @param startTime Start time in ms.
     * @param endTime   End time in ms.
     * @param limit     Results limit.
     *
     * @return The request to execute.
     */
    @JvmOverloads
    fun getAllOrders(symbol: String, orderId: Long? = null, startTime: Long? = null, endTime: Long? = null, limit: Int? = null) =
        Request(service.getAllOrders(symbol, orderId, startTime, endTime, limit))

    /**
     * Send in an OCO order.
     *
     * @param symbol                 Symbol the order is made on.
     * @param side                   Order side (BUY/SELL).
     * @param quantity               Order quantity.
     * @param price                  Order price.
     * @param stopPrice              Stop price.
     * @param stopLimitPrice         Stop limit price.
     * @param stopLimitTimeInForce   Valid values are GTC/FOK/IOC.
     * @param listClientOrderId      A unique Id for the entire orderList
     * @param limitClientOrderId     A unique Id for the limit order
     * @param limitIcebergQuantity   Used to make the LIMIT_MAKER leg an iceberg order.
     * @param stopClientOrderId      A unique id for the stop loss/stop loss limit leg
     * @param stopIcebergQuantity    Used with STOP_LOSS_LIMIT leg to make an iceberg order
     * @param limitStrategyId        Limit strategy id.
     * @param limitStrategyType      The value cannot be less than 1000000.
     * @param trailingDelta          Trailing delta.
     * @param stopStrategyId         Stop strategy id.
     * @param stopStrategyType       The value cannot be less than 1000000.
     *
     * @return The request to execute.
     */
    @JvmOverloads
    fun newOCO(
        symbol: String,
        side: OrderSide,
        quantity: String,
        price: String,
        stopPrice: String,
        stopLimitPrice: String?,
        stopLimitTimeInForce: TimeInForce? = null,
        newOrderRespType: NewOrderResponseType? = null,
        listClientOrderId: String? = null,
        limitClientOrderId: String? = null,
        limitIcebergQuantity: String? = null,
        stopClientOrderId: String? = null,
        stopIcebergQuantity: String? = null,
        limitStrategyId: Int? = null,
        limitStrategyType: Int? = null,
        trailingDelta: Long? = null,
        stopStrategyId: Int? = null,
        stopStrategyType: Int? = null
    ) = Request(
        service.newOCO(
            symbol,
            side,
            quantity,
            price,
            stopPrice,
            stopLimitPrice,
            stopLimitTimeInForce,
            newOrderRespType,
            listClientOrderId,
            limitClientOrderId,
            limitIcebergQuantity,
            stopClientOrderId,
            stopIcebergQuantity,
            limitStrategyId,
            limitStrategyType,
            trailingDelta,
            stopStrategyId,
            stopStrategyType
        )
    )

    /**
     * Cancel an entire Order List. Canceling an individual leg will cancel the
     * entire OCO
     *
     * @param symbol            Symbol to cancel the order.
     * @param orderListId       Either `orderListId` or `listClientOrderId` must be provided.
     * @param listClientOrderId Either `orderListId` or `listClientOrderId` must be provided.
     * @param newClientOrderId  Used to uniquely identify this cancel. Automatically generated by default.
     *
     * @return The request to execute.
     */
    @JvmOverloads
    fun cancelOCO(symbol: String, orderListId: Long? = null, listClientOrderId: String? = null, newClientOrderId: String? = null) =
        Request(service.cancelOCO(symbol, orderListId, listClientOrderId, newClientOrderId))

    /**
     * Retrieves a specific OCO based on provided optional parameters.
     *
     * @param orderListId       Either `orderListId` or `origClientOrderId` must be provided.
     * @param origClientOrderId Either `orderListId` or `origClientOrderId` must be provided.
     *
     * @return The request to execute.
     */
    fun getOCO(orderListId: Long? = null, origClientOrderId: String? = null) = Request(service.getOCO(orderListId, origClientOrderId))

    /**
     * Retrieves all OCO based on provided optional parameters.
     *
     * @param fromId    ID to search from.
     * @param startTime Start time in ms.
     * @param endTime   End time in ms.
     * @param limit     Results limit.
     *
     * @return The request to execute.
     */
    @JvmOverloads
    fun getAllOCO(fromId: Long? = null, startTime: Long? = null, endTime: Long? = null, limit: Int? = null) =
        Request(service.getAllOCO(fromId, startTime, endTime, limit))

    /**
     * Retrieves all open OCO.
     *
     * @return The request to execute.
     */
    fun getOpenOCO() = Request(service.getOpenOCO())

    /**
     * Get current account information.
     *
     * @return The request to execute.
     */
    fun getAccount() = Request(service.getAccount())

    /**
     * Get trades for a specific account and symbol. If fromId is set, it will get
     * id &gt;= fromId. Otherwise, most recent trades are returned.
     *
     *
     * @param symbol    Symbol of the trade.
     * @param orderId   Order id of the trade.
     * @param fromId    ID from which we search trades.
     * @param startTime Start time in ms.
     * @param endTime   End time in ms.
     * @param limit     Results limit.
     *
     * @return The request to execute.
     */
    @JvmOverloads
    fun getTrades(symbol: String, orderId: Long? = null, fromId: Long? = null, startTime: Long? = null, endTime: Long? = null, limit: Int? = null) =
        Request(service.getTrades(symbol, orderId, fromId, startTime, endTime, limit))

    /**
     * Displays the user's current order count usage for all intervals with default
     * request
     *
     * @return The request to execute.
     */
    fun getOrderCount() = Request(service.getOrderCount())
}